## Task 1

In [1]:
import random
import numpy as np

def check_board(alice_nums, bob_nums, row, col):
    # Function to check if Alice and Bob win a game
    # given their numbers, row and column
    
    # Check if binary sum of Alice's numbers == 0
    if (sum(alice_nums) % 2) != 0:
        return False
    
    # Check if binary sum of Bob's numbers == 1
    if (sum(bob_nums) % 2) != 1:
        return False
    
    # Check if Alice and Bob placed same number in cell where her row and his column intersect
    if alice_nums[col] != bob_nums[row]:
        return False
    
    return True


# print(check_board([1, 0, 1], [1, 1, 0], 1, 2))
# print(check_board([0, 0, 0], [0, 0, 1], 2, 2))
# print(check_board([0, 1, 1], [1, 0, 0], 0, 1))

In [2]:
def simulate_task1():
    '''
    returns True if Alice and Bob won
            False if Alice and Bob lost
    '''
    
    
    # Richard randomly chooses row and column
    row = random.randint(0, 2)
    col = random.randint(0, 2)
    
    # the values filled are represented by the binary representation of the random int
    rand1 = random.randint(0, 7)
    rand2 = random.randint(0, 7)
    
    alice_nums = [int((rand1 & (1 << x)) > 0) for x in range(3)]
    bob_nums = [int((rand2 & (1 << x)) > 0) for x in range(3)]
    
    return check_board(alice_nums, bob_nums, row, col)


results = [simulate_task1() for _ in range(500000)]

print('Task 1')
print(f'{sum(results)} games won')
print(f'{len(results)} games played')
print(f'{sum(results)/len(results):.5f} of games won')

Task 1
62466 games won
500000 games played
0.12493 of games won


In [3]:
def generate_all_boards_task1():    
    all_boards = []
    for row in range(3):
        for col in range(3):
            for rand1 in range(8):
                for rand2 in range(8):
                    alice_nums = [int((rand1 & (1 << x)) > 0) for x in range(3)]
                    bob_nums = [int((rand2 & (1 << x)) > 0) for x in range(3)]
                    all_boards.append((alice_nums, bob_nums, row, col))
    return all_boards

results = [check_board(*board) for board in generate_all_boards_task1()]

print('Task 1')
print(f'{sum(results)} winning boards')
print(f'{len(results)} total possible boards')
print(f'{sum(results)/len(results):.5f} probability of winning')

Task 1
72 winning boards
576 total possible boards
0.12500 probability of winning


Through simulation of all possible states of the board, we can determine that the probability of winning would be $\frac{72}{576}=0.125$. This is rather close to the value obtained from above.

## Task 2

We make a minor adjustment to the solution for Task 1. Instead of randomly generating all $3$ bits, we only generate the first $2$ bits. The first $2$ bits will determine what the last bit is in order to satisfy the binary sum condition.

Alice:

| Bit 1 | Bit 2 | Bit 3 |
|:-:|:-:|:-:|
| 0 | 0 | 0 |
| 0 | 1 | 1 |
| 1 | 0 | 1 |
| 1 | 1 | 0 |

Bob:

| Bit 1 | Bit 2 | Bit 3 |
|:-:|:-:|:-:|
| 0 | 0 | 1 |
| 0 | 1 | 0 |
| 1 | 0 | 0 |
| 1 | 1 | 1 |

In [4]:
def simulate_task2():
    '''
    returns True if Alice and Bob won
            False if Alice and Bob lost
    '''
    
    # Mostly similar to Task 1
    
    row = random.randint(0, 2)
    col = random.randint(0, 2)
    
    # Only generate first 2 bits
    # Last bit will be uniquely determined by the first 2 bits
    rand1 = random.randint(0, 3)
    rand2 = random.randint(0, 3)
    
    alice_nums = [int((rand1 & (1 << x)) > 0) for x in range(2)]
    alice_nums.append((2 - sum(alice_nums)) % 2)
    bob_nums = [int((rand2 & (1 << x)) > 0) for x in range(2)]
    bob_nums.append((3 - sum(bob_nums)) % 2)
    
    return check_board(alice_nums, bob_nums, row, col)


results = [simulate_task2() for _ in range(500000)]

print('Task 2')
print(f'{sum(results)} games won')
print(f'{len(results)} games played')
print(f'{sum(results)/len(results):.5f} of games won')

Task 2
249704 games won
500000 games played
0.49941 of games won


In [5]:
def generate_all_boards_task2():    
    all_boards = []
    for row in range(3):
        for col in range(3):
            for rand1 in range(4):
                for rand2 in range(4):
                    alice_nums = [int((rand1 & (1 << x)) > 0) for x in range(2)]
                    alice_nums.append((2 - sum(alice_nums)) % 2)
                    bob_nums = [int((rand2 & (1 << x)) > 0) for x in range(2)]
                    bob_nums.append((3 - sum(bob_nums)) % 2)
                    all_boards.append((alice_nums, bob_nums, row, col))
    return all_boards

results = [check_board(*board) for board in generate_all_boards_task2()]

print('Task 2')
print(f'{sum(results)} winning boards')
print(f'{len(results)} total possible boards')
print(f'{sum(results)/len(results):.5f} probability of winning')

Task 2
72 winning boards
144 total possible boards
0.50000 probability of winning


## Task 3

Possible strategy:
Alice always puts $[0, 1, 1]$. If Bob is told first column, Bob puts $[1, 0, 0]$. Otherwise, Bob puts $[1, 1, 1]$. 

Using this strategy there are only $9$ possible boards, which are determined by the row and column given by Richard. They will lose only if Richard says row $1$, column $1$.

Hence this strategy will give $\frac{8}{9}$ winning probability, which is the most optimal.

In [6]:
def simulate_task3():
    '''
    returns True if Alice and Bob won
            False if Alice and Bob lost
    '''
    
    row = random.randint(0, 2)
    col = random.randint(0, 2)
    
    alice_nums = [0, 1, 1]
    bob_nums = [1, 1, 1]
    
    if col == 0:
        bob_nums = [1, 0, 0]
    
    return check_board(alice_nums, bob_nums, row, col)


results = [simulate_task3() for _ in range(500000)]

print('Task 3')
print(f'{sum(results)} games won')
print(f'{len(results)} games played')
print(f'{sum(results)/len(results):.5f} of games won')

Task 3
444406 games won
500000 games played
0.88881 of games won


In [7]:
def generate_all_boards_task3():    
    all_boards = []
    for row in range(3):
        for col in range(3):
            alice_nums = [0, 1, 1]
            bob_nums = [1, 1, 1]

            if col == 0:
                bob_nums = [1, 0, 0]
            all_boards.append((alice_nums, bob_nums, row, col))
    return all_boards

results = [check_board(*board) for board in generate_all_boards_task3()]

print('Task 3')
print(f'{sum(results)} winning boards')
print(f'{len(results)} total possible boards')
print(f'{sum(results)/len(results):.5f} probability of winning')

Task 3
8 winning boards
9 total possible boards
0.88889 probability of winning


## Task 4

In [8]:
%matplotlib inline
# Importing standard Qiskit libraries
import math
import qiskit as q
from qiskit import QuantumCircuit, execute, Aer, IBMQ
from qiskit.compiler import transpile, assemble
from qiskit.tools.jupyter import *
from qiskit.visualization import *
from ibm_quantum_widgets import *

# Loading your IBM Q account(s)
provider = IBMQ.load_account()
backend = Aer.get_backend('qasm_simulator')


$$\newcommand{\ket}[1]{\left|{#1}\right\rangle}$$
Alice is $q_0$ and $q_1$, while Bob is given $q_2$ and $q_3$. The qubits are entangled by $\ket{q_0 q_2}=\ket{q_1 q_3}=\ket{\Phi^+}=CNOT(H \otimes I)\ket{00}$.

The following table is taken from [Reference B](http://electron6.phys.utk.edu/phys250/modules/module%203/a_pseudotelepathy_game.htm).

| | col 1 | col 2 | col 3 |
|:-:|:-:|:-:|:-:|
| row 1 | $I \otimes S_z$ | $S_z \otimes I$ | $-S_z \otimes S_z$ |
| row 2 | $S_x \otimes I$ | $I \otimes S_x$ | $-S_x \otimes S_x$ |
| row 3 | $S_x \otimes S_z$ | $S_z \otimes S_x$ | $-S_y \otimes S_y$ |

Each cell in this table shows the basis which they should measure their qubits in. The result of the measurment is what they will fill in the cell. We will perform measurements to find the values in the first and second cell of each row and column.

### Bob's strategy

If Bob is given column $1$, he needs to measure $q_3$ in the $S_z$ basis and $q_2$ in the $S_x$ basis. The cell in row $1$ will take the measurement of $q_3$, and row $2$ will take the measurement of $q_2$. This can be implemented using a $H$ gate on $q_2$ followed by a $SWAP$ gate on $q_2$ and $q_3$.

If Bob is given column $2$, he needs to measure $q_2$ in the $S_z$ basis and $q_3$ in the $S_x$ basis. The cell in row $1$ will take the measurement of $q_2$, and row $2$ will take the measurement of $q_3$. This can be implemented by using a $H$ gate on $q_2$ only.

If Bob is given column $3$, the result of measururing $-S_z \otimes S_z$ goes into row $1$, and result of measuring $-S_x \otimes S_x$ goes into row $2$. This can be done using a $CNOT$ gate with $q_3$ as the control and $q_2$ as the target, followed by a $H$ gate on $q_2$.

### Alice's strategy

If Alice is given row $1$, she needs to measure both $q_0$ and $q_1$ in the $S_z$ basis. The cell in column $1$ will take the measurement of $q_1$, and column $2$ will take the measurement of $q_0$. This can be implemented using a $SWAP$ gate only.

If Alice is given row $2$, she needs to measure both $q_0$ and $q_1$ in the $S_x$ basis. The cell in column $1$ will take the measurement of $q_0$, and column $2$ will take the measurement of $q_1$. This can be simply implemented by using $H$ gates on both $q_0$ and $q_1$.

If Alice is given row $3$, the result of measururing $S_x \otimes S_z$ goes into column $1$, and result of measuring $S_z \otimes S_x$ goes into column $2$. This can be done using a $H$ gate on $q_1$, followed by $X$ gates on both $q_0$ and $q_1$, followed by $CNOT$ gate with $q_1$ as the control and $q_0$ as the target, followed by a $H$ gate on $q_0$.



In [9]:
def alice_task4(circ, row):
    if row == 0:
        circ.swap(0, 1)
    if row == 1:
        circ.h(0)
        circ.h(1)
    if row == 2:
        circ.x(0)
        circ.x(1)
        circ.h(0)
        circ.cx(1, 0)
        circ.h(1)

def bob_task4(circ, col):
    if col == 0:
        circ.h(2)
        circ.swap(2, 3)
    if col == 1:
        circ.h(3)
    if col == 2:
        circ.cx(3, 2)
        circ.h(3)

In [10]:
def simulate_task4():
    '''
    returns True if Alice and Bob won
            False if Alice and Bob lost
    '''
    
    row = random.randint(0, 2)
    col = random.randint(0, 2)
    
    # Prepare entangled states
    circ = q.QuantumCircuit(4, 4)
    circ.h(0)
    circ.h(1)
    circ.cx(0, 2)
    circ.cx(1, 3)

    alice_task4(circ, row)
    bob_task4(circ, col)
    
    circ.measure(0, 0)
    circ.measure(1, 1)
    circ.measure(2, 2)
    circ.measure(3, 3)

    job = execute(circ, backend, shots=1, memory=True)
    alice_nums = [int(job.result().get_memory()[0][3]), int(job.result().get_memory()[0][2])]
    alice_nums.append((2 - sum(alice_nums)) % 2)
    bob_nums = [int(job.result().get_memory()[0][1]), int(job.result().get_memory()[0][0])]
    bob_nums.append((3 - sum(bob_nums)) % 2)
    
    return check_board(alice_nums, bob_nums, row, col)

# Simulating circuits is slow, hence number of runs has been reduced
results = [simulate_task4() for _ in range(5000)]

print('Task 4')
print(f'{sum(results)} games won')
print(f'{len(results)} games played')
print(f'{sum(results)/len(results):.5f} of games won')

Task 4
5000 games won
5000 games played
1.00000 of games won


In [11]:
def generate_all_boards_task4():    
    all_boards = []
    for row in range(3):
        for col in range(3):
            
            # Prepare entangled states
            circ = q.QuantumCircuit(4, 4)
            circ.h(0)
            circ.h(1)
            circ.cx(0, 2)
            circ.cx(1, 3)
            
            alice_task4(circ, row)
            bob_task4(circ, col)
            
            circ.measure(0, 0)
            circ.measure(1, 1)
            circ.measure(2, 2)
            circ.measure(3, 3)
            
            job = execute(circ, backend, shots=1, memory=True)
            alice_nums = [int(job.result().get_memory()[0][3]), int(job.result().get_memory()[0][2])]
            alice_nums.append((2 - sum(alice_nums)) % 2)
            bob_nums = [int(job.result().get_memory()[0][1]), int(job.result().get_memory()[0][0])]
            bob_nums.append((3 - sum(bob_nums)) % 2)
            
            all_boards.append((alice_nums, bob_nums, row, col))
    return all_boards


results = [check_board(*board) for board in generate_all_boards_task4()]

print('Task 4')
print(f'{sum(results)} winning boards')
print(f'{len(results)} total possible boards')
print(f'{sum(results)/len(results):.5f} probability of winning')

Task 4
9 winning boards
9 total possible boards
1.00000 probability of winning


The simulations indeed show that the quantum strategy can allow for a $100\%$ win rate.